Backend/services/user/index.js

```diff
@@ -1,44 +1,150 @@
+        });
+      }
+      response.status(200).json(result.rows);
+    }
+  );
+});
 
 /**
  * Returns the user with the given id.
  */
-router.get('/:id', (request, response) => {
-    response.status(500).send({
-        error: 'Pending Implementations'
-    })
-})
+router.get("/:id", auth.authenticateToken, (request, response) => {
+  console.log(auth.authenticateToken);
+  const id = parseInt(request.params.id);
+  pool.query(
+    "SELECT ID, users.firstname, users.lastname, users.username FROM users WHERE id = $1",
+    [id],
+    (error, result) => {
+      if (error) {
+        response.status(500).send({
+          error: error,
+        });
+      }
+      if (result.rows[0].username === request.user.username) {
+        return response.status(200).json(result.rows[0]);
+      }
+      return response.status(200).send("not allowed");
+    }
+  );
+});
 
 /**
  * Registers a new user.
  */
-router.post('/register', (request, response) => {
+router.post("/register", async (request, response) => {
+  const { firstName, lastName, username, password } = request.body;
+
+  try {
+    const salt = await bcrypt.genSalt();
+    const hashedPassword = await bcrypt.hash(password, salt);
+    if (username) {
+      pool.query(
+        `SELECT users.username FROM users WHERE users.username = $1`,
+        [username],
+        function (error, result) {
+          if (error) {
+            response.status(500).send({
+              error: error,
+            });
+          }
+          if (result.rows.length < 1) {
+            pool.query(
+              "INSERT INTO users (firstname, lastname, username, password) VALUES ($1, $2, $3, $4) RETURNING id",
+              [firstName, lastName, username, hashedPassword],
+              (error, result) => {
+                if (error) {
+                  response.status(500).send({
+                    error: error,
+                  });
+                }
+                auth.generateAccessToken(request, response, result);
+              }
+            );
+          } else {
+            response.send("Username already taken");
+          }
+        }
+      );
+    }
+  } catch (error) {
     response.status(500).send({
-        error: 'Pending Implementations'
-    })
-})
+      error: error,
+    });
+  }
+});
 
 /**
  * Authenticates a user.
  */
-router.post('/login', (request, response) => {
-    response.status(500).send({
-        error: 'Pending Implementations'
-    })
-})
+router.post("/login", async (request, response) => {
+  const { username, password } = request.body;
+
+  if (username) {
+    pool.query(
+      `SELECT users.username, users.password, users.id FROM users WHERE users.username = $1`,
+      [username],
+      async function (error, result) {
+        if (error) {
+          response.status(500).send({
+            error: error,
+          });
+        }
+        if (result.rowCount === 0) {
+          return response.status(400).send("Cannot find user");
```

Comment:
`400` status code is for bad request.

Use `404` for resources that aren't found. Since this context is authentication, it may also be relevant to use `401`.